### PR TITLE
ROCKETMQ-82: RocketMQ-Flink Integration

### PR DESCRIPTION
**Pull request by vesense**
# What is the purpose of the change
RocketMQ integration for [Apache Flink](https://flink.apache.org/). This module includes the RocketMQ source and sink that allows a flink job to either write messages into a topic or read from topics in a flink job.
https://issues.apache.org/jira/browse/ROCKETMQ-82

# Brief changelog
* RocketMQSource - The RocketMQSource is based on RocketMQ pull consumer mode, and provides exactly once reliability guarantees when checkpoints are enabled.
Otherwise, the source doesn't provide any reliability guarantees.
* RocketMQSink - The RocketMQSink provides at-least-once reliability guarantees when checkpoints are enabled and `withBatchFlushOnCheckpoint(true)` is set.
Otherwise, the sink reliability guarantees depends on rocketmq producer's retry policy, for this case, the messages sending way is sync by default,
but you can change it by invoking `withAsync(true)`. 
* KeyValueDeserializationSchema - The main API for deserializing topic and tags is the `org.apache.rocketmq.flink.common.serialization.KeyValueDeserializationSchema` interface.
`rocketmq-flink` includes general purpose `KeyValueDeserializationSchema` implementations called `SimpleKeyValueDeserializationSchema`.
* KeyValueSerializationSchema - The main API for serializing topic and tags is the `org.apache.rocketmq.flink.common.serialization.KeyValueSerializationSchema` interface.
`rocketmq-flink` includes general purpose `KeyValueSerializationSchema` implementations called `SimpleKeyValueSerializationSchema`.
* TopicSelector - The main API for selecting topic and tags is the `org.apache.rocketmq.flink.common.selector.TopicSelector` interface.
`rocketmq-flink` includes general purpose `TopicSelector` implementations called `DefaultTopicSelector` and `SimpleTopicSelector`.
* RocketMQFlinkExample - which receive messages from RocketMQ brokers and send messages to broker after processing.

# Verifying this change

local & unit tests check passed ✅ 
apache-rat check passed ✅ 
checkstyle check passed ✅

### CODE REVIEW

**Comment by vesense** [comment]
@vongosling @zhouxinyu Do yo have time to take a look?

**Comment by vesense** [comment]
Any comments are welcome..

**Comment by vesense** [comment]
Will add more unit tests later.

**Comment by vesense** [comment]
Added unit tests.

**Comment by zhouxinyu** [comment]
Cool, let's move this PR forward.

**Comment by vesense** [comment]
I'm glad to see that some users have used this patch in some scenes. But I would suggest to fork the code from official repo, not my personal.
And I will always keep my eyes on the RocketMQ streaming integration.

**Comment by vesense** [comment]
Also, I will update the status of integrating and SQL modules in community mailing list periodically.

File: rocketmq-flink/README.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+| NAME        | DESCRIPTION           | DEFAULT  |
+| ------------- |:-------------:|:------:|
+| nameserver.address      | name server address *Required* | null |
+| nameserver.poll.interval      | name server poll topic info interval     |   30000 |
+| brokerserver.heartbeat.interval | broker server heartbeat interval      |    30000 |
+| consumer.group | consumer group *Required*     |    null |
+| consumer.topic | consumer topic *Required*       |    null |
+| consumer.tag | consumer topic tag      |    * |
+| consumer.offset.reset.to | what to do when there is no initial offset on the server      |   latest/earliest/timestamp |
+| consumer.offset.from.timestamp | the timestamp when `consumer.offset.reset.to=timestamp` was set   |   $TIMESTAMP |
```
Comment (zhouxinyu) [inline rocketmq-flink/README.md]: The default value $TIMESTAMP and $UUID are confusing, especially the $ symbol, a short description may be better.

File: rocketmq-flink/pom.xml [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    <groupId>org.apache.rocketmq</groupId>
+    <artifactId>rocketmq-flink</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <packaging>jar</packaging>
+
+    <developers>
+        <developer>
+            <id>vesense</id>
+            <name>Xin Wang</name>
+            <email>xinwang@apache.org</email>
```
Comment (zhouxinyu) [inline rocketmq-flink/pom.xml]: Not recommended to use personal user info, see: https://github.com/apache/rocketmq/blob/master/pom.xml

File: rocketmq-flink/pom.xml [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        </developer>
+    </developers>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <!--maven properties -->
+        <maven.test.skip>true</maven.test.skip>
+        <maven.javadoc.skip>true</maven.javadoc.skip>
```
Comment (zhouxinyu) [inline rocketmq-flink/pom.xml]: IMO, it's not a good practice that skip test and javadoc.

File: rocketmq-flink/pom.xml [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            <groupId>org.apache.rocketmq</groupId>
+            <artifactId>rocketmq-namesrv</artifactId>
+            <version>${rocketmq.version}</version>
+            <scope>test</scope>
+            <!--<exclusions>-->
+                <!--<exclusion>-->
+                    <!--<groupId>ch.qos.logback</groupId>-->
+                    <!--<artifactId>logback-classic</artifactId>-->
+                <!--</exclusion>-->
+            <!--</exclusions>-->
```
Comment (zhouxinyu) [inline rocketmq-flink/pom.xml]: If the exclusions section isn't necessary, just remove it and make code clean.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMqConfig.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import org.apache.rocketmq.client.consumer.DefaultMQPullConsumer;
+import org.apache.rocketmq.client.producer.DefaultMQProducer;
+import org.apache.rocketmq.common.protocol.heartbeat.MessageModel;
+
+import static org.apache.rocketmq.flink.RocketMqUtils.getInteger;
+
+/**
+ * RocketMqConfig for Consumer/Producer.
+ */
+public class RocketMqConfig {
```
Comment (zhouxinyu) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMqConfig.java]: Use `RocketMQConfig`

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            return;
+        }
+
+        if (async) {
+            // async sending
+            try {
+                producer.send(msg, new SendCallback() {
+                    @Override
+                    public void onSuccess(SendResult sendResult) {
+                        LOG.debug("Async send message success!");
```
Comment (zhouxinyu) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java]: how about appending the message id in send result to debug?

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                    }
+                });
+            } catch (Exception e) {
+                LOG.error("Async send message failure!", e);
+            }
+        } else {
+            // sync sending, will return a SendResult
+            try {
+                SendResult result = producer.send(msg);
+                if (LOG.isDebugEnabled()) {
```
Comment (zhouxinyu) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java]: No need to check whether debug is enabled.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+                            List<MessageExt> messages = pullResult.getMsgFoundList();
+                            for (MessageExt msg : messages) {
+                                byte[] key = msg.getKeys() != null ? msg.getKeys().getBytes(StandardCharsets.UTF_8) : null;
+                                byte[] value = msg.getBody();
+                                OUT data = schema.deserializeKeyAndValue(key, value);
+
+                                // output and state update are atomic
+                                synchronized (lock) {
+                                    context.collectWithTimestamp(data, msg.getBornTimestamp());
+                                    putMessageQueueOffset(mq, pullResult.getNextBeginOffset());
```
Comment (zhouxinyu) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: Do we need use the next begin offset in other PullStatus?

File: rocketmq-flink/README.md (hunk 1 above)
Comment (vesense) [reply rocketmq-flink/README.md]: Good catch. Will fix.

File: rocketmq-flink/pom.xml (hunk 2 above)
Comment (vesense) [reply rocketmq-flink/pom.xml]: Will remove.

File: rocketmq-flink/pom.xml (hunk 3 above)
Comment (vesense) [reply rocketmq-flink/pom.xml]: Will fix.

File: rocketmq-flink/pom.xml (hunk 4 above)
Comment (vesense) [reply rocketmq-flink/pom.xml]: Will remove.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java (hunk 6 above)
Comment (vesense) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java]: I think we can print the `SendResult ` here following the way of sync sending.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java (hunk 7 above)
Comment (vesense) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java]: Yes. Will remove.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java (hunk 8 above)
Comment (vesense) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: Yes. We need the `nextBeginOffset` in the `PullResult` to update the offset table which will be checkpointing when `snapshotState` invoked .

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMqConfig.java (hunk 5 above)
Comment (vesense) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMqConfig.java]: Will fix.

Comment (vesense) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMqConfig.java]: Also `RocketMqUtils` -> `RocketMQUtils`.

**Comment by vesense** [comment]
@zhouxinyu Addressed your comments.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java (hunk 8 above)
Comment (zhouxinyu) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: I mean we also should update the offset when the pullresult is OFFSET_ILLEGAL or NO_MATCHED_MSG

**Comment by vesense** [comment]
@zhouxinyu Addressed all of your comments. Please take a look again.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void invoke(IN input, Context context) throws Exception {
+        Message msg = prepareMessage(input);
+
+        if (batchFlushOnCheckpoint) {
+            batchList.add(msg);
```
Comment (zhouxinyu) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java]: Add action isn't protected by the lock `batchList`， is this a concurrent issue?

Comment (vesense) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java]: Not an issue. Each RocketMQSink instance belong to a thread. We just need to sync `batchList ` when `snapshotState` invoked.

Comment (zhouxinyu) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java]: Got it.

**Comment by zhouxinyu** [comment]
LGTM now~

**Comment by vesense** [comment]
Thanks @zhouxinyu

**Comment by vesense** [comment]
@zhouxinyu Can we merge this in?

**Comment by vongosling** [comment]
LGTM

**Pull request merged**